### DISCUSSION TITLE
Installation Experiences

### DISCUSSION THREAD

**Discussion by marcjamali** [spine-tools/SpineOpt.jl#374] [Support]
Hello,
Here is some information about my installation experience of Spine Toolbox, SpineOpt and other software like Python & Julia.  I'm sorry if this is a little long.

I used the procedure or parts of the procedure written on https://spine-project.github.io/SpineOpt.jl/latest/getting_started/installation/.  Everything below was executed starting from a clean Windows 10 pc.
__________________________________________________________________
**1st try** :

Started by installing Spine-Toolbox using spine-toolbox-0.6.0-final.0-x64.exe.  Then trying to install Julia from the File –> Settings –> Tools\Install Julia button.  The result in the console is

> Julia installation started
> $ C:\Program Files\Spine Toolbox\tools\python.exe -m jill install --confirm --install_dir C:\Users\Moi\AppData\Local\julias --symlink_dir C:\Users\Moi\AppData\Local\julias\bin
> Fatal Python error: initfsencoding: unable to load the file system codec
> Traceback (most recent call last):
> 	Process crashed
> 
> Julia installation failed

I then installed Python 3.8.10 but got the same result.

2nd try:
I tried installing Python first, then Spine Toolbox but got the same result as above.

3rd try:
I tried installing Python first by making sure I check the "Add Python 3.8 to PATH" checkbox in the installation process.  Then I installed the Spine Toolbox.  Then, when trying to install Julia from the File –> Settings –> Tools\Install Julia button.  The result in the console was different:
Result in the console is

> Julia installation started
> $ C:\Users\Moi\AppData\Local\Programs\Python\Python38\python.exe -m jill install --confirm --install_dir C:\Users\Moi\AppData\Local\julias --symlink_dir C:\Users\Moi\AppData\Local\julias\bin
> C:\Users\Moi\AppData\Local\Programs\Python\Python38\python.exe: No module named jill
> 
> Julia installation failed
> 

I then installed Julia 1.6.1 manually but Julia is not in the "Julia kernel spec..." selection list in File –> Settings –> Tools and when going in the Kernel Specification Editor, things don't look good
![Kernel Spec Editor](https://user-images.githubusercontent.com/81410782/117685204-fdc9f480-b1b5-11eb-853b-839209a05476.jpg)

4th try:
Last try.  I install Python making sure it's added to PATH.  Same with Julia 1.6.1.  Then Spine-Toolbox.  Julia 1.6 doesn't appear in the Kernel spec selection list but the path to my julia.exe file appears in light grey underneath and "Using Julia default project" also (Were they there in my previous trials? I don't know).

I then press on the "Add/Update SpineOpt" button and the installation of SpineOpt is successfull and from there SpineOpt works.

Conclusion and remarks:
It works but if we follow the instructions line by line in the Getting Started guide of SpineOpt and SpineToolbox installation chapters, there are some steps where we're not sure:

- do I install Python or not?  The Requirements chapter is written after the Installation chapter.
- do I use the official release or the development version of Spine-Toolbox (The link to the installation coming from SpineOpt brings to the development version installation chapter (https://github.com/Spine-project/Spine-Toolbox#installation).  I used the official release.
- the step starting with "Select a Julia Kernel spec. If there is none, you ....".  This is the step that bothered me the most.  See all trials above.  Results in consoles look like something is wrong even when SpineOpt works (4th try).

If you need more information on what I did or did not do, I'll be glad to help.

Marc

**Comment by DillonJ** [comment]
Thanks for this @marcjamali 

What do you think @manuelma @mihlema @soininen? It seems like we could make the getting started guide clearer and tidy up the whole whole Julia configuration toolbox side. I think we discussed this already and the work @manuelma should make this easier regarding default consoles etc. - we just need to make sure the documentation aligns.

**Reply by manuelma** [reply]
The 'easier' console is not part of release-0.6. It's only in the development version at the moment (latest master). I guess at some point we can write the steps to just use this new console and see if that helps.

Some comments on @marcjamali issues:

> Fatal Python error: initfsencoding: unable to load the file system codec (1st try)

It looks like the Python embedded in spinetoolbox installer has a problem. We weren't aware of this, but maybe @PekkaSavolainen can take a look.

> C:\Users\Moi\AppData\Local\Programs\Python\Python38\python.exe: No module named jill (3rd try)

This is normal, as the python you're using doesn't have the package jill. I believe `pip install jill` would solve this.

> in the Kernel Specification Editor, things don't look good (3rd try)

Did you press the button that says 'Make kernel specification'? Kernel specs aren't created automatically and thus do not appear in the list (until you press that button and the process goes well). Another option is to just accept the Settings without opening the Kernel Editor, and if you have selected 'Use embedded console", the necessary kernel spec is automatically created at that point.

> "Select a Julia Kernel spec. If there is none, you ....". This is the step that bothered me the most

I hear you, this is really annoying for new users and I have experienced it many times. The jupyter kernel requires too much configuration and is not really a mandatory requirement. In the new development version we're trying to make it optional by providing a non-jupyter environment that can run spine opt nevertheless.

**Reply by marcjamali** [reply]
Concerning the jill package:  I installed python 3.8 downloaded from the python website.  I guess that if jill is not included in the default python installation, it could be a step in the installation process of spine.

Concerning the Kernel: To tell you the truth, I have no idea what this is for.  I would need a training on what a kernel is and why we need that and what does it eat in winter.  I would therefore need a very detailed step-by-step installation process for that part if it is important.  When I said, things don't look good (with the screenshot I pasted in the 3rd try), you can imagine me going in the kernel editor, not knowing what I'm doing and seeing a red row written python in it while I'm trying to do something right for Julia.  In those cases, I press on the "Cancel" button. So, no, I didn't press on the "Make Kernel Specification".
